metagpt/roles/di/data_interpreter.py

```diff
@@ -72,14 +136,49 @@ async def _write_and_exec_code(self, max_retry: int = 3):
                 if ReviewConst.CHANGE_WORDS[0] in review:
                     counter = 0  # redo the task again with help of human suggestions
 
-        return code["code"], result, success
+        return code, result, success
 
-    async def _write_code(self):
-        todo = WriteCodeWithoutTools() if not self.use_tools else WriteCodeWithTools(selected_tools=self.tools)
+    async def _write_code(
+        self,
+        counter,
+        plan_status="",
+        tool_info="",
+    ):
+        todo = WriteAnalysisCode()
         logger.info(f"ready to {todo.name}")
+        use_reflection = counter > 0 and self.use_reflection
+
+        user_requirement = self.get_memories()[0].content
 
-        context = self.planner.get_useful_memories()
-        # print(*context, sep="\n***\n")
-        code = await todo.run(context=context, plan=self.planner.plan, temperature=0.0)
+        code = await todo.run(
+            user_requirement=user_requirement,
+            plan_status=plan_status,
+            tool_info=tool_info,
+            working_memory=self.working_memory.get(),
+            use_reflection=use_reflection,
+        )
 
         return code, todo
+
+    async def _check_data(self):
+        if (
+            not self.use_plan
+            or not self.planner.plan.get_finished_tasks()
+            or self.planner.plan.current_task.task_type
+            not in [
+                TaskType.DATA_PREPROCESS.type_name,
+                TaskType.FEATURE_ENGINEERING.type_name,
+                TaskType.MODEL_TRAIN.type_name,
+            ]
+        ):
+            return
+        logger.info("Check updated data")
+        code = await CheckData().run(self.planner.plan)
```

Comment:
Pass in the necessary parameters when instantiating the action to complete the correct priority setting of the context / config

Reply:
Corrected WriteAnalysisCode. For this action, since we want rc.todo to be continuously WriteAnalysisCode, it is a bit inconvenient to set todo to CheckData and back to WriteAnalysisCode